enigma-principal/app/src/epoch_u/epoch_provider.rs

```diff
@@ -56,146 +54,255 @@ impl EpochProvider {
+        let guard = self.lock_guard_or_wait()?;
+        if guard.is_empty() {
+            mem::drop(guard);
+            return Ok(false);
+        }
+        let last = guard.iter().last();
+        //TODO: why borrow checker fails here
+//        mem::drop(guard);
+        let epoch_state = match last {
+            Some(epoch_state) => epoch_state,
+            None => {
+                return Err(EpochStateUndefinedErr {}.into());
+            }
+        };
+        let is_unconfirmed = match &epoch_state.confirmed_state {
+            Some(_) => false,
+            None => true,
         };
-        mem::drop(guard);
-        Ok(epoch_state)
+        Ok(is_unconfirmed)
     }
 
-    /// Reset the `EpochState` stores in memory
-    pub fn reset_epoch_state(&self) -> Result<(), Error> {
-        self.set_or_clear_epoch_state(None)
+    /// Return a list of all confirmed `EpochState`
+    pub fn get_all_confirmed(&self) -> Result<Vec<EpochState>, Error> {
+        let guard = self.lock_guard_or_wait()?;
+        let mut result: Vec<EpochState> = vec![];
+        for epoch_state in guard.iter() {
+            if epoch_state.confirmed_state.is_some() {
+                result.push(epoch_state.clone());
+            }
+        }
+        Ok(result)
     }
 
-    fn set_epoch_state(&self, epoch_state: EpochState) -> Result<(), Error> {
-        self.set_or_clear_epoch_state(Some(epoch_state))
+    /// Returns the confirmed `EpochState` for the epoch of the block number
+    /// # Arguments
+    ///
+    /// * `block_number` - A block number in the desired epoch
+    pub fn get_confirmed_by_block_number(&self, block_number: U256) -> Result<EpochState, Error> {
+        let mut result: Option<&EpochState> = None;
+        let epoch_states = self.get_all_confirmed()?;
+        for epoch_state in epoch_states.iter() {
+            let confirmed: &ConfirmedEpochState = epoch_state.confirmed_state.as_ref().unwrap();
+            if confirmed.block_number <= block_number {
+                result = Some(epoch_state);
+            }
+        }
+        match result {
+            Some(epoch_state) => Ok(epoch_state.clone()),
+            None => {
+                Err(EpochStateTransitionErr { current_state: EPOCH_STATE_UNCONFIRMED.to_string() }.into())
+            }
+        }
+    }
+
+    /// Returns the most recent `EpochState` stored in memory
+    /// # Arguments
+    ///
+    /// * `exclude_unconfirmed` - Exclude any unconfirmed state
+    pub fn last(&self, exclude_unconfirmed: bool) -> Result<EpochState, Error> {
+        let guard = self.lock_guard_or_wait()?;
+        let mut epoch_state_val: Option<EpochState> = None;
+        for epoch_state in guard.iter().rev() {
+            if (exclude_unconfirmed && epoch_state.confirmed_state.is_some()) || !exclude_unconfirmed {
+                epoch_state_val = Some(epoch_state.clone());
+                break;
+            }
+        }
+        mem::drop(guard);
+        match epoch_state_val {
```

Comment:
you can do `epoch_state_val.ok_or(EpochStateUndefinedErr{})`
(you might need to add `into()` on the error or do `Ok(epoch_state_val.ok_or(EpochStateUndefinedErr{})?)` which will do it for you)

Reply:
Fixed fa9ed2c39c3c56a8c1fc52c2269aab08e3f75487